### PR TITLE
of-precice: add new versions

### PR DESCRIPTION
**Pull request by kjrstory**
Of-precice was not maintained, except for old commits. I've registered a new version and made it work properly.
I tested with openfoam@2306, precice@2.5.0, and of-precice@1.2.3 on Linux Ubuntu 20.04 with the Cascade Lake architecture and GCC version 9.4.0.

### CODE REVIEW

File: var/spack/repos/builtin/packages/of-precice/package.py
Comment (MakisH) [inline var/spack/repos/builtin/packages/of-precice/package.py:19]:
```suggestion
    url = "https://github.com/precice/openfoam-adapter/archive/v1.2.3.tar.gz"
    maintainers("MakisH", "kjrstory")
```

**Review by MakisH — COMMENTED** [review]
I am a maintainer of the upstream [openfoam-precice adapter](https://github.com/precice/openfoam-adapter) project and co-maintainer of the [preCICE Spack package](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/precice/package.py), and I found this PR by chance. I see that the package currently has no maintainers (original author: @olesenm) and I am happy to be a maintainer, together with @kjrstory. I added a suggestion on how to declare that.

All the changes look good and important. The checksums are correct, the new dependency is indeed a dependency. The `tutorials` folder does not exist anymore (previously only on the `master`/`develop` branches when the package was first created). Adding `of-precice/lib` to the `LD_LIBRARY_PATH` seems to be important to be able to correctly load the library at runtime. I am not sure if this is the best way, but many other packages seem to be following the same approach.

P.S.: In the long run, I would either rename this package to better reflect the purpose and be discoverable (e.g., `openfoam-with-precice`) or, better, merge this as an optional feature to the `openfoam` package, as these two are tightly connected.

**Comment by kjrstory** [comment]
Hello! Nice to meet you. @MakisH 
 
I'm actively involved as a maintainer for Openfoam-org in Spack. With a growing interest in Fluid-Structure Interaction, I've extended my focus to Precice and have been contributing to the package recipe for of-precice(openfoam-precice-adapter) in Spack. Unfortunately, of-precice has been lacking maintenance for quite some time. Moreover, it currently only supports the openfoam.com version and doesn't extend support to the openfoam.org version, despite the fact that it is supported upstream. There are plans to address these issues and improve them in the future. Also, I want to support adapters beyond OpenFOAM in Spack.

**Comment by MakisH** [comment]
> Moreover, it currently only supports the openfoam.com version and doesn't extend support to the openfoam.org version, despite the fact that it is supported upstream.

Yet another reason to make the adapter as an optional feature of each OpenFOAM[.com/.org] Spack recipe. There are already multiple recipes for OpenFOAM, so making additional packages for each OpenFOAM that provide the preCICE adapter would bring additional complexity.

In any case, the changes that this PR brings are very welcome. If you apply my suggestion (setting maintainers), I can approve, and this may be enough to get this merged.

Of course, if you do not agree with my suggestion above, we can also discuss alternatives.

**Review by MakisH — APPROVED** [review]

**Comment by MakisH** [comment]
@alalazo I was assigned this issue, but I don't have enough rights to resolve the conflicts and merge.

File: var/spack/repos/builtin/packages/of-precice/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/of-precice/package.py:99]: Do you know why setting `LD_LIBRARY_PATH` is needed at runtime?

Comment (MakisH) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: As I wrote above:

> Adding `of-precice/lib` to the LD_LIBRARY_PATH seems to be important to be able to correctly load the library at runtime. I am not sure if this is the best way, but many other packages seem to be following the same approach.

I don't know why @kjrstory had to specify that.

Comment (MakisH) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: Maybe as context: This is a shared library that OpenFOAM will have to load at runtime.

Comment (alalazo) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: Is the library loaded with `dlopen` ? In that case `LD_LIBRARY_PATH` is not needed :thinking:

Comment (MakisH) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: OpenFOAM does seem to use `dlopen` to load dynamic libraries, but I don't know if this also applies to libraries loaded by function objects (it should): https://develop.openfoam.com/search?group_id=23&nav_source=navbar&project_id=47&repository_ref=master&search=dlopen&search_code=true

Comment (alalazo) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: @kjrstory Do you recall why you added these 2 lines? Spack usually sets RPATHs for link dependencies, so this shouldn't be needed.

Comment (olesenm) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: > @kjrstory Do you recall why you added these 2 lines? Spack usually sets RPATHs for link dependencies, so this shouldn't be needed.

Except that this precice adaptor is being _"bolted on"_ to an OpenFOAM compilation, so the OpenFOAM rpath will not (cannot) include it.

Comment (MakisH) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: What do you mean? Is this something we do wrong in the adapter that we could (should!) fix?

**Comment by alalazo** [comment]
There is a conflict to be solved with develop

File: var/spack/repos/builtin/packages/of-precice/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: @olesenm Apologies, but I have no real knowledge of OpenFOAM or its extensions. In general though:
- If this package installs applications / libraries that are used by potential dependents, then RPATH should suffice
- If this package installs a library that is a plugin of one of its dependencies (i.e. if it is dlopened by the dependency), then `LD_LIBRARY_PATH` doesn't need to be set - the path to the library should be sufficient

Am I missing something?

Comment (olesenm) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: No, nothing for you to do. The point being that while OpenFOAM can be built with rpath for its various dependencies (eg, scotch, mpi etc), but precice is being build afterwards and thus its install location (for rpath use) cannot be known when OpenFOAM is being built.  The LD_LIBRARY_PATH remains the only feasible way inform OpenFOAM where/how it should find arbitrary additional libraries to load.

Comment (MakisH) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: We essentially have the following dependency situation:

```mermaid
graph TD;
   of-precice-->precice;
   of-precice-->openfoam;
   openfoam-->of-precice;
```

OpenFOAM loads the `of-precice` library, which also depends on preCICE. The `of-precice` library also depends on OpenFOAM libraries, but this is not the problem here.

preCICE itself should be discoverable at runtime. Package recipe: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/precice/package.py (cc @fsimonis)

Comment (MakisH) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: Without setting `LD_LIBRARY_PATH` the concrete error is:

```
--> FOAM Warning : 
    From void* Foam::dlLibraryTable::openLibrary(const Foam::fileName&, bool)
    in file db/dynamicLibrary/dlLibraryTable/dlLibraryTable.C at line 188
    Could not load "libpreciceAdapterFunctionObject.so"
libpreciceAdapterFunctionObject.so: cannot open shared object file: No such file or directory
--> FOAM Warning : 
Unknown function type preciceAdapterFunctionObject

Valid function types :

11
(
BilgerMixtureFraction
ObukhovLength
patchProbes
probes
psiReactionThermoMoleFractions
psiSpecieReactionRates
rhoReactionThermoMoleFractions
rhoSpecieReactionRates
setTimeStepFaRegion
sets
surfaces
)



    From static Foam::autoPtr<Foam::functionObject> Foam::functionObject::New(const Foam::word&, const Foam::Time&, const Foam::dictionary&)
    in file db/functionObjects/functionObject/functionObject.C at line 129.
--> loading function object 'preCICE_Adapter'
```

Note that the execution of OpenFOAM will continue, so someone that does not know what output to expect may interepret the exit code as normal.

Here is my complete `LD_LIBRARY_PATH`. Setting it here adds the `of-precice-1.2.3-<hash>/lib` and moves around the `libfabric` (no clue why). OpenFOAM itself also adds paths to `LD_LIBRARY_PATH` ([package recipe](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/openfoam/package.py)).

```diff
$ echo $LD_LIBRARY_PATH 
+ /home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/of-precice-1.2.3-e5qkh6ehxgawrwjokcm6r2echlskg5eu/lib:
+ /home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib64:
+/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib:
+/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib64:
+/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib:
/home/gc/OpenFOAM/gc-v2306/platforms/linux64GccDPInt32-spack/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/site/2306/platforms/linux64GccDPInt32-spack/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/platforms/linux64GccDPInt32-spack/lib/user-mpi:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/platforms/linux64GccDPInt32-spack/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/boost-1.83.0-eilj2osypasejhg7tqzqkoznuqxlea3r/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/cgal-4.13-cvcnwlf37jg52lxs4e5omkl7bfpzwy42/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/adios2-2.9.2-gsym5co4c2grkffzwiugdrvoghk4tfmx/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/platforms/linux64GccDPInt32-spack/lib/dummy:
/home/gc/OpenFOAM/gc-v2306/platforms/linux64GccDPInt32-spack/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/site/2306/platforms/linux64GccDPInt32-spack/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/platforms/linux64GccDPInt32-spack/lib/user-mpi:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/platforms/linux64GccDPInt32-spack/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/boost-1.83.0-eilj2osypasejhg7tqzqkoznuqxlea3r/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/cgal-4.13-cvcnwlf37jg52lxs4e5omkl7bfpzwy42/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/adios2-2.9.2-gsym5co4c2grkffzwiugdrvoghk4tfmx/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/openfoam-2306-j3sqxbvqmosq3zuzdmdfbaryucvigqev/platforms/linux64GccDPInt32-spack/lib/dummy:
- /home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib64:
- /home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib:
- /home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib64:
- /home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/libfabric-1.20.0-n2sff4drcg3lsjhahkox3uwts547kkei/lib:
/home/gc/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/hdf5-1.14.3-qn5aemrswjz4xggv3xqq4zppnbalhmhi/lib

```

Comment (alalazo) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: @olesenm This is mostly for my understanding, but:

> The LD_LIBRARY_PATH remains the only feasible way inform OpenFOAM where/how it should find arbitrary additional libraries to load.

As far as I know it is not the only feasible way. If a random library is loaded at runtime, the application needs to know its location. On linux this means going through a:
```c
void *dlopen(const char *filename, int flags);
```
call. `LD_LIBRARY_PATH` enters into that only if filename is not a "pathname" (i.e. it doesn't contain a `/` char). 

Is that the case in OpenFOAM? Aren't there other mechanisms in place to `dlopen` stuff using a pathname (e.g. read from an OpenFOAM specific environment variables a list of dirs where to find plugins)?

Comment (olesenm) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: This is true, but would mean that someone would have to specify something like this in their input file:
```
libs
(
    "/home/spack/linux-opensuse_leap15-haswell/gcc-7.3.1/of-precice-VER-hutl75kvogp3l3ekpexqxcfi2vvjlpyc
/libName.so"
);
```
Looks non-trivial to write.  The other idea, which you allude to would be to have an environment variable for plugins.

**Review by alalazo — CHANGES_REQUESTED** [review]
LGTM, once the conflict is fixed

File: var/spack/repos/builtin/packages/of-precice/package.py
Comment (fsimonis) [reply var/spack/repos/builtin/packages/of-precice/package.py:99]: Other preCICE maintainer here. 

Doesn't spack already extend LD_LIBRARY_PATH when loading a package?
Loading both openfoam and the  adapter together should currently work.
`spack load openfoam of-precice`

A pragmatic solution to require only the adapter to be loaded could be to declare openfoam also as a run dependency.

When `spack load`ing the adapter package, spack adds the adapter /lib to LD_LIBRARY_PATH and adds openfoam to PATH due to the run dependency.

**Comment by alalazo** [comment]
Why was this closed?

**Comment by kjrstory** [comment]
> Why was this closed?

@alalazo
I apologize. While I was modifying the commit, I accidentally closed this pull request. I am currently resolving the conflict and will commit the changes again and reopening this pull request.

**Comment by alalazo** [comment]
@kjrstory No worries, I was just wondering if there was an issue I wasn't aware of because this PR had to be closed, or if it was superseded by another PR :slightly_smiling_face:

**Comment by MakisH** [comment]
@kjrstory I guess this can be reopened and stay open, since it only needs resolving some conflicts, no? Otherwise, I would open a new one eventually. (I don't have the rights to reopen)

**Review by alalazo — APPROVED** [review]

**Pull request merged**
